docs/CONTRIBUTING.md

```diff
@@ -35,3 +35,47 @@ Test is determined fragile/broken when it is failing despite no defects are pres
   - Test can still be resurrected from history if needed - in ATH or plugin itself.
 
 The situation is expected to be handled in 90 days since reported. In case of no corrective action, the test will be removed from ATH.
+
+## Guidelines to avoid fragile/flaky tests
+
+### Click and wait
+
+When writing acceptance tests using Selenium it is common to click on an element and then run some other action on the resulting page.
+Most of the time the page is updated really quick (specially if there is not a network request involved), but on slow infrastructure this can take some time so that the next test action after clicking could be checking on something which is not the expected.
+
+To avoid this, use the click and wait pattern:
+
+```
+WebElement clickableElement = ...;
+clickableElement.click();
+waitFor(by.id("my-element")).withTimeout(10);
+... now run actions safely on my-element ...
+```
+
+This way the test will not be affected by an underlying slow infrastructure.
+
+### Do not rely on external infra
+
+When the test requires to interact with an external system, use a `DockerFixture`.
+There are multiple examples into the [`fixtures` directory](https://github.com/jenkinsci/acceptance-test-harness/tree/master/src/main/java/org/jenkinsci/test/acceptance/docker/fixtures).
+
+### Do not use sleep
+
+Sleeping blindly is a bad practice, it just makes tests execution time longer. 
+Use the [`Wait` class](https://github.com/jenkinsci/acceptance-test-harness/blob/master/src/main/java/org/jenkinsci/test/acceptance/junit/Wait.java) as much as possible. `CapybaraPortingLayerImpl#waitFor` is an easy way to create instances of `Wait`.
+
+### Use robust selectors
+
+When selecting elements from the DOM, use robust selectors. 
+In general, try to avoid selectors that are known to change or are outside your control (they are not in your code). 
+
+A few examples:
```

Comment:
It would make sense to add a few links to good examples from our project.

Reply:
And maybe also mention the [form-element-path-plugin](https://wiki.jenkins.io/display/JENKINS/Form+Element+Path+Plugin)

Reply:
> It would make sense to add a few links to good examples from our project.

I'm trying to avoid linking sources. Code evolves and links get obsolete quickly, specially if the link is to a specific line of code. We could use commit links, which are granted to stay the same, but then we would be linking to old sources quite quickly.

> And maybe also mention the form-element-path-plugin

It's mentioned. See bullet 3.